be/src/exec/vectorized/sorting/sort_helper.h

```diff
@@ -0,0 +1,160 @@
+    const NullData& null_data = column->immutable_null_column_data();
+    auto null_pred = [&](const SmallPermuteItem& item) -> bool {
+        if (is_null_first) {
+            return null_data[item.index_in_chunk] == 1;
+        } else {
+            return null_data[item.index_in_chunk] != 1;
+        }
+    };
+
+    VLOG(2) << fmt::format("nullable column tie before sort: {}\n", fmt::join(tie, ","));
+    VLOG(2) << fmt::format("nullable column before sort: {}\n", dubug_column(column, permutation));
+
+    TieIterator iterator(tie, range.first, range.second);
+    while (iterator.next()) {
+        int range_first = iterator.range_first;
+        int range_last = iterator.range_last;
+
+        if (range_last - range_first > 1) {
+            auto pivot_iter =
+                    std::partition(permutation.begin() + range_first, permutation.begin() + range_last, null_pred);
+            int pivot_start = pivot_iter - permutation.begin();
+            int notnull_start = is_null_first ? pivot_start : range_first;
+            int notnull_end = is_null_first ? range_last : pivot_start;
+
+            if (notnull_start < notnull_end) {
+                tie[pivot_start] = 0;
+                RETURN_IF_ERROR(sort_and_tie_column(cancel, column->data_column(), is_asc_order, is_null_first,
+                                                    permutation, tie, {notnull_start, notnull_end}, build_tie));
+            }
+        }
+
+        VLOG(3) << fmt::format("column after iteration: [{}, {}): {}\n", range_first, range_last,
+                               dubug_column(column, permutation));
+        VLOG(3) << fmt::format("tie after iteration: [{}, {}] {}\n", range_first, range_last, fmt::join(tie, ",    "));
+    }
+
+    VLOG(2) << fmt::format("nullable column tie after sort: {}\n", fmt::join(tie, ",    "));
+    VLOG(2) << fmt::format("nullable column after sort: {}\n", dubug_column(column, permutation));
+
+    return Status::OK();
+}
+
+template <class DataComparator, class PermutationType>
+static inline Status sort_and_tie_helper(const bool& cancel, const Column* column, bool is_asc_order,
+                                         PermutationType& permutation, Tie& tie, DataComparator cmp,
+                                         std::pair<int, int> range, bool build_tie) {
+    auto lesser = [&](auto lhs, auto rhs) { return cmp(lhs, rhs) < 0; };
+    auto greater = [&](auto lhs, auto rhs) { return cmp(lhs, rhs) > 0; };
+    auto do_sort = [&](auto begin, auto end) {
+        if (is_asc_order) {
+            ::pdqsort(cancel, begin, end, lesser);
+        } else {
+            ::pdqsort(cancel, begin, end, greater);
+        }
+    };
+
+    VLOG(2) << fmt::format("tie before sort: {}\n", fmt::join(tie, ","));
+    VLOG(2) << fmt::format("column before sort: {}\n", dubug_column(column, permutation));
+
+    TieIterator iterator(tie, range.first, range.second);
+    while (iterator.next()) {
```

Comment:
may be we need check cancel flag here? otherwise may be we couldn't cancel a large sort query.